ledger/participant-integration-api/src/main/scala/platform/store/dao/CommandCompletionsReader.scala

```diff
@@ -31,24 +33,40 @@ private[dao] final class CommandCompletionsReader(
   )(implicit
       loggingContext: LoggingContext
   ): Source[(Offset, CompletionStreamResponse), NotUsed] = {
-    Source
-      .future(
-        dispatcher
-          .executeSql(metrics.daml.index.db.getCompletions) { implicit connection =>
-            queryNonPruned.executeSql[List[CompletionStreamResponse]](
-              query = storageBackend.commandCompletions(
-                startExclusive = startExclusive,
-                endInclusive = endInclusive,
-                applicationId = applicationId,
-                parties = parties,
-              )(connection),
-              minOffsetExclusive = startExclusive,
-              error = pruned =>
-                s"Command completions request from ${startExclusive.toHexString} to ${endInclusive.toHexString} overlaps with pruned offset ${pruned.toHexString}",
-            )
-          }
-      )
-      .mapConcat(_.map(response => offsetFor(response) -> response))
+    val pruneSafeQuery =
+      (range: QueryRange[Offset]) => { implicit connection: Connection =>
+        queryNonPruned.executeSql[Vector[CompletionStreamResponse]](
+          query = storageBackend.commandCompletions(
+            startExclusive = range.startExclusive,
+            endInclusive = range.endInclusive,
+            applicationId = applicationId,
+            parties = parties,
+            limit = pageSize,
+          )(connection),
+          minOffsetExclusive = startExclusive,
+          error = (prunedUpTo: Offset) =>
+            s"Command completions request from ${startExclusive.toHexString} to ${endInclusive.toHexString} overlaps with pruned offset ${prunedUpTo.toHexString}",
+        )
+      }
+
+    val initialRange = new QueryRange[Offset](
+      startExclusive = startExclusive,
+      endInclusive = endInclusive,
+    )
+    val source: Source[CompletionStreamResponse, NotUsed] = PaginatingAsyncStream
+      .streamFromSeekPagination[QueryRange[Offset], CompletionStreamResponse](
+        startFromOffset = initialRange,
+        getOffset = (previousCompletion: CompletionStreamResponse) => {
+          val lastOffset = offsetFor(previousCompletion)
+          initialRange.copy(startExclusive = lastOffset)
+        },
+      ) { subRange: QueryRange[Offset] =>
+        dispatcher.executeSql(metrics.daml.index.db.getCompletions)(pruneSafeQuery(subRange))
+      }
+    source.map(response => offsetFor(response) -> response)
   }
 
+  private def offsetFor(response: CompletionStreamResponse): Offset =
+    ApiOffset.assertFromString(response.checkpoint.get.offset.get.getAbsolute)
```

Comment:
This code preexisted, so no need to change it now. Though, it strikes me as brittle. It depends on existence of a field that is declared as optional on the grpc interface. Maybe leave a TODO to consider enforcing offset existence by construction in the result of `commandCompletions`.

Reply:
Move this method to it's previous location and added a comment (though without a 'TODO')